polys: remove the ring cache

oscarbenjamin opened the pull request:
Rings and fields like `QQ[x]` were cached in a cache that grows without bound. This commit removes the cache altogether and changes places that depend on the cache. The main need for the cache is that each ring dynamically creates a class for its elements and then uses `isinstance` to check if an object is an element of the ring. Here we change all places that use `isinstance` to compare the rings instead since the rings already have `__eq__` defined.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fix for #24581
See also #20857 and #23690

#### Other comments

This is not yet ready for merge. I will restore some sort of ring cache but one that does not grow without bound.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
   * [WIP]: The global ring cache for rings like `QQ[x]` has been changed...
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
polys: remove the ring cache

Rings and fields like QQ[x] were cached in a cache that grows without
bound. This commit removes the cache altogether and changes places that
depend on the cache. The main need for the cache is that each ring
dynamically creates a class for its elements and then uses isinstance to
check if an object is an element of the ring. Here we change all places
that use isinstance to compare the rings instead since the rings already
have __eq__ defined.

A subsequent commit will restore some sort of ring cache but one that
does not grow without bound.

oscarbenjamin committed:
polys: remove unneeded import

asmeurer commented on sympy/polys/fields.py:
```diff
@@ -493,9 +489,9 @@ def __mul__(f, g):
+        elif isinstance(g, FracElement) and g.field == field:
             return f.new(f.numer*g.numer, f.denom*g.denom)
-        elif isinstance(g, field.ring.dtype):
+        elif isinstance(g, PolyElement) and g.ring == field.ring:
```
Shouldn't the `isinstance(g, PolyElement)` be built-in to the `__eq__` of `g.ring`?

oscarbenjamin replied:
If `g` is not a `PolyElement` then it might not have a `ring` attribute.

asmeurer commented:
I'm happy to see this if you can get it working. I have no idea what this would break, both in terms of the semantics of the dynamic class creation and the performance, but if you keep a handle on those, or are at least confident that the tests cover enough, then this should be OK.

I'm also not sure to what degree we can break compatibility with this code. Is it public API? I'd love to get rid of the dynamic class creation. Not only does it make the code harder to maintain and lead to subtle issues like this, but it also breaks a lot of stuff like pickling (and I wouldn't be surprised if it actually makes the performance worse).

oscarbenjamin committed:
polys: override of_type in QQ[x] and QQ(x)

Since polynomial rings are not singletons we need to specialise the
of_type method for PolynomialRing and FractionField.

oscarbenjamin commented:
> I'm also not sure to what degree we can break compatibility with this code. Is it public API?

I think it is sort of public. It's not widely used but it is definitely used. The question is how deeply though i.e. the `dtype` attribute was not previously documented until I added this:
https://docs.sympy.org/latest/modules/polys/domainsref.html#sympy.polys.domains.domain.Domain.dtype
I think it can just be changed though. The dynamic class creation is a bad idea and should be changed. Very little code will depend on this.

oscarbenjamin commented:
> Significantly changed benchmark results (PR vs master)
> 
> ```diff
>        before           after         ratio
>      [7f676531]       [73429adc]
> +      29.3±0.3ms       58.4±0.8ms     1.99  integrate.TimeIntegrationRisch01.time_doit(1)
> +        38.0±1ms       67.7±0.9ms     1.78  integrate.TimeIntegrationRisch01.time_doit(10)
> +      29.3±0.4ms       57.9±0.7ms     1.98  integrate.TimeIntegrationRisch01.time_doit_risch(1)
> +      39.1±0.9ms         68.1±1ms     1.74  integrate.TimeIntegrationRisch01.time_doit_risch(10)
> +         205±2ms          323±4ms     1.57  integrate.TimeIntegrationRisch03.time_doit(10)
> +         198±3ms          310±2ms     1.57  integrate.TimeIntegrationRisch03.time_doit_risch(10)
> +        85.3±1ms          167±3ms     1.95  solve.TimeMatrixSolvePyDySlow.time_linsolve(1)
> +      83.4±0.6ms          168±2ms     2.01  solve.TimeMatrixSolvePyDySlow.time_solve(1)
> +        502±20μs      1.11±0.02ms     2.20  solve.TimeRationalSystemSymbol.time_linsolve(1)
> +      28.5±0.6ms          232±8ms     8.14  solve.TimeSolve01.time_solve
> +      28.5±0.4ms          230±3ms     8.05  solve.TimeSolve01.time_solve_nocheck
> +        38.9±1ms          167±5ms     4.29  solve.TimeSolveLinSys189x49.time_solve_lin_sys
> ```

Looks like completely removing the cache is not great for performance. More investigation needed...

oscarbenjamin commented:
> Looks like completely removing the cache is not great for performance.

Two tangential remarks:

1. It's good that the CI benchmarks can pick up on this which suggests that they do do their job at least on a basic level.
2. At the same time it's hard to interpret this because timings in the presence of a cache are always tricky: it isn't clear how these timing differences relate to  real situations because caches always look good when something is repeated in a loop. This is not an issue with the CI benchmarks but with tight-loop benchmarks in general.

oscarbenjamin commented:
The benchmarks that have slowed down are those that involve the EX domain. In the EX domain elements are Expr and are simplified using `cancel`. This ends up repeatedly creating the same polynomial rings. For example the `TimeSolve01.time_solve` benchmark is these linear equations:
```python
In [9]: c_0, c_1, c_2, c_3 = syms = symbols('c_0:4')

In [10]: x0, y0, x1, y0_0, y0_1, y1_0, y1_1 = symbols('x0, y0, x1, y0_0, y0_1, y1_0, y1_1')

In [11]: eqs = [c_0 + c_1*x0 + c_2*x0**2 + c_3*x0**3 - y0_0, c_0 + c_1*x1 + c_2*x1**2 + c_3*x1**3 - y1_0,
    ...:  c_1 + 2*c_2*x0 + 3*c_3*x0**2 - y0_1, c_1 + 2*c_2*x1 + 3*c_3*x1**2 - y1_1]

In [12]: eqs = [c_0 + c_1*x0 + c_2*x0**2 + c_3*x0**3 - y0_0, c_0 + c_1*x1 + c_2*x1**2 + c_3*x1**3 - y1_0,
    ...:  c_1 + 2*c_2*x0 + 3*c_3*x0**2 - y0_1, c_1 + 2*c_2*x1 + 3*c_3*x1**2 - y1_1]

In [13]: c_0, c_1, c_2, c_3 = syms = symbols('c_0:4')

In [14]: x0, y0, x1, y0_0, y0_1, y1_0, y1_1 = symbols('x0, y0, x1, y0_0, y0_1, y1_0, y1_1')

In [15]: eqs = [c_0 + c_1*x0 + c_2*x0**2 + c_3*x0**3 - y0_0, c_0 + c_1*x1 + c_2*x1**2 + c_3*x1**3 - y1_0,
    ...:  c_1 + 2*c_2*x0 + 3*c_3*x0**2 - y0_1, c_1 + 2*c_2*x1 + 3*c_3*x1**2 - y1_1]

In [16]: eqs
Out[16]: 
⎡                  2        3                           2        3                               2     
⎣c₀ + c₁⋅x₀ + c₂⋅x₀  + c₃⋅x₀  - y₀ ₀, c₀ + c₁⋅x₁ + c₂⋅x₁  + c₃⋅x₁  - y₁ ₀, c₁ + 2⋅c₂⋅x₀ + 3⋅c₃⋅x₀  - y₀

                          2       ⎤
 ₁, c₁ + 2⋅c₂⋅x₁ + 3⋅c₃⋅x₁  - y₁ ₁⎦

In [17]: eqs[0]
Out[17]: 
                  2        3       
c₀ + c₁⋅x₀ + c₂⋅x₀  + c₃⋅x₀  - y₀ ₀

In [18]: eqs[1]
Out[18]: 
                  2        3       
c₀ + c₁⋅x₁ + c₂⋅x₁  + c₃⋅x₁  - y₁ ₀

In [19]: syms
Out[19]: (c₀, c₁, c₂, c₃)

In [20]: %time solve(eqs, syms)
```
The following rings are created:
```
ZZ[x0, y0_0]
ZZ[x1, y1_0]
ZZ[x0, y0_1]
ZZ[x1, y1_1]
ZZ[x0, x1, y0_0, y0_1, y1_0, y1_1]
ZZ(x0,x1,y0_0,y0_1,y1_0,y1_1)[c_0, c_1, c_2, c_3]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]ZZ[x0, y0_0]
ZZ[x1, y1_0]
ZZ[x0, y0_1]
ZZ[x1, y1_1]
ZZ[x0, x1, y0_0, y0_1, y1_0, y1_1]
ZZ(x0,x1,y0_0,y0_1,y1_0,y1_1)[c_0, c_1, c_2, c_3]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[x1, y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_0, y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y0_1, y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_0, y1_1]
ZZ[y1_1]
ZZ[y1_1]
```
What this shows is that even with the cache we can end up creating a lot of different rings in one operation. I suppose that the first thing to do is try to optimise the cost of constructing a ring and then we can think about how exactly to cache them. Also the memory usage of the rings should probably be minimised so that caching is less costly.

oscarbenjamin committed:
polys: cache the PolyRing.clone method

The PolyRing.clone method is used to generate rings similar to an existing
polynomial ring e.g. by PlyRing.drop. Since this ends up being used in the
inner loop of various operations like heugcd we cache the result of the clone
method so that as long as this ring is in memory it can keep references to any
related cloned rings without needing to rebuild them from scratch. This could
be an alternative to the global ring cache that was previously removed.

oscarbenjamin commented:
I've pushed another commit that reduces the time taken to construct a polynomial ring by using a cache for the MonomialOps class. Profiling revealed that to be the slowest part. Now the cost of constructing a ring is not so much different from the cost of retrieving one from a cache.

oscarbenjamin committed:
polys: cache MonomialOps codegen routines

Caches the most expensive part of constructing a polynomial ring which is
generating and compiling the code for monomial operations in the MonomialOps
class. First we have a LRU cache for MonomialOps instances of a given number
of generators. Then we also cache the compiled methods generated by each
MonomialOps instance.

oscarbenjamin commented:
The benchmarks no longer show any significant difference after the last commit. There are some more changes to be considered though.

Firstly this is not a backwards incompatible change. This changes the meaning of the dtype attribute in some sense. Previously `dtype` was the type of all elements of a ring. Here dtype still functions as a constructor for elements but it is not guaranteed that all elements are instances of dtype rather than of an equivalent class. Any code expecting `isinstance(obj, dtype)` to work as it did before might fail.

However it is not clear to me why any code should need to do use `isinstance` like that. The places I needed to change here are all just methods like `__add__` that check so that they can raise `TypeError` if the user attempts to add elements of different rings.

I think that it should be okay to make this change although we would need to decide how to document this:
https://docs.sympy.org/latest/modules/polys/domainsref.html#sympy.polys.domains.domain.Domain.dtype

There are some other things to tidy up here and also other potential improvements to reduce memory consumption and to make rings with large numbers of generators more efficient in terms of space and time (#20857).

oscarbenjamin commented:
> This changes the meaning of the dtype attribute in some sense

It might be better to replace `dtype` with a function so that it can still convert inputs but will give an explicit error is someone tries to use with isinstance:
```python
In [2]: isinstance(1, lambda : None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-37a0752e4fc5> in <module>
----> 1 isinstance(1, lambda : None)

TypeError: isinstance() arg 2 must be a type or tuple of types
```
Then there is no risk of silently getting the wrong result.

oscarbenjamin committed:
polys: remove redundant code paths

oscarbenjamin committed:
polys: add is_element to PolyRing and FracField

oscarbenjamin committed:
polys: temporarily skip doctest

oscarbenjamin committed:
polys: remove dynamic classes in PolyRing and FracField

oscarbenjamin commented:
The last commit removes the dynamically created classes altogether.

oscarbenjamin commented:
Currently this removes the dynamic class creation from `PolyRing` and `FracField`. Without having a dynamic class we have a situation where a `PolyElement` and `FracElement` need to take an additional constructor argument to specify the associated ring or field. That means that the constructors take a new signature but that is unavoidable if we don't have the dynamic classes.

The question now then is what `dtype` should mean then. Currently on master ring.dtype is a constructor that construct elements and is also the type of the elements and that type is unique to the particular ring/field. Making this unique type work as intended when the rings or fields might be created independently is achieved through a cache. Here in this PR we remove that cache and therefore the unique type cannot be used.

The constructor looks like:
```python
In [1]: dtype = ring('x y', ZZ)[0].dtype

In [2]: dtype({(1,2):ZZ(3)})
Out[2]: 3*x*y**2
```
The fact that it is the type means that:
```python
In [3]: zero = dtype()

In [4]: isinstance(zero, dtype)
Out[4]: True
```
Lastly on master checking `isinstance` like this guarantees that something is an element of this particular ring or field.

The change currently in this PR is that `dtype` is now a callable function but is not the type of the elements:
```python
In [5]: dtype({(1,2):ZZ(3)})
Out[5]: 3*x*y**2

In [6]: dtype({(1,2):ZZ(3)})
Out[6]: 3*x*y**2

In [7]: zero = dtype({})

In [8]: isinstance(zero, dtype)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-a3cb57c106f9> in <cell line: 1>()
----> 1 isinstance(zero, dtype)

TypeError: isinstance() arg 2 must be a type or tuple of types
```

Removing the cache and removing the dynamically generated classes cannot be done without some change of behaviour here because it isn't possible for the `dtype` attribute to function exactly as it does on master without the dynamic classes and the cache. As I see it there are two possible options:

1. Have dtype be a constructor that still functions the same way as before.
2. Have dtype be the type but not be a constructor any more.

Keeping dtype as a type does not completely preserve the current `isinstance` behaviour because the type would just have to be the `PolyElement` class and it would be the same type for all elements of all rings. That means that using dtype with isinstance will not guarantee that an object is an element of *this* ring any more. Since it isn't possible to preserve that behaviour with its current meaning I opted to keep `dtype` as a constructor because that behaviour can be preserved correctly. This way using `isinstance` with `dtype` will raise an exception but that is better than getting a wrong result because the meaning of `isinstance` has quietly changed.

I have added an alternative method `is_element` to `PolyRing` and `FracField` which can be used as e.g. `R.is_element(x)` to check if `x` is an element of the ring `R`. This is something that we could potentially extend more generally to all domains. Looking at the type handling and conversions across the domain system and PolyElement and FracField I expect that a lot of things could be simplified a bit and made faster by using `is_element` rather than `isinstance`.

With the PR as it stands I would need to update the docs at least for `dtype` but I just want to see what anyone thinks about the backwards compatibility part of these changes.

Does anyone have any views on what the appropriate choice for `dtype` should be?
